### PR TITLE
Add Dashboard

### PR DESCRIPTION
**Pull request by zaelgohary**
Related Issues:
- https://github.com/codescalers/CSHR/issues/272
- https://github.com/codescalers/CSHR/issues/302

image

image

image

image

image

image

### CODE REVIEW

File: client/src/components/AddOffice.vue [hunk 1]
```diff
@@ -0,0 +1,80 @@
...
+    const selectedCountry = ref(countryList[0])
+    const office = ref({
+      name: '',
+      country: selectedCountry.value,
+      weekend: selectedWeekend.value.value
+    })
+
+    async function addOffice() {
+      await $api.office.create(office.value)
+    }
```
Comment (MohamedElmdary) [inline client/src/components/AddOffice.vue]: Handle error and loading statuses

Tip: use `useAsyncState` from `@vueuse/core`

File: client/src/components/AddUser.vue [hunk 2]
```diff
@@ -0,0 +1,338 @@
...
+    const birthdayPicker = ref(false)
+    const joiningDatePicker = ref(false)
+    const toggleDatePicker = (picker: string) => {
+      if (picker === 'birthdayPicker') birthdayPicker.value = !birthdayPicker.value
+      else if (picker === 'joiningDatePicker') joiningDatePicker.value = !joiningDatePicker.value
+    }
+
+    async function addUser() {
+      await $api.auth.register(user.value as Api.Inputs.Register)
+    }
```
Comment (MohamedElmdary) [inline client/src/components/AddUser.vue]: Handle Error and Loading statuses

File: client/src/components/AddUser.vue [hunk 3]
```diff
@@ -0,0 +1,338 @@
...
+      offices.value = (await $api.office.list()).map((office: any) => ({
+        id: office.id,
+        name: office.name
+      }))
+      location.value = offices.value[0]
+      supervisors.value = ((await $api.users.admin.list()) as any)
+        .filter((supervisor: any) => supervisor.user_type === 'Supervisor')
+        .map((supervisor: any) => ({ id: supervisor.id, name: supervisor.full_name }))
+      selectedSupervisor.value = supervisors.value[0]
+    })
```
Comment (MohamedElmdary) [inline client/src/components/AddUser.vue]: Handle Error and Loading statuses

File: client/src/components/SetUserVacations.vue [hunk 4]
```diff
@@ -0,0 +1,107 @@
...
+      annual_leaves: 0,
+      leave_excuses: 0,
+      emergency_leaves: 0
+    })
+
+    watchEffect(async () => {
+      if (selectedUser.value) {
+        const result = await $api.vacations.balance.list({
+          user_ids: selectedUser.value.id
+        })
```
Comment (MohamedElmdary) [inline client/src/components/SetUserVacations.vue]: Handle Error and Loading statuses

File: client/src/components/SetUserVacations.vue [hunk 5]
```diff
@@ -0,0 +1,107 @@
...
+          const { annual_leaves, leave_excuses, emergency_leaves } = userVacations.value
+          vacation.value.annual_leaves = annual_leaves.all
+          vacation.value.leave_excuses = leave_excuses.all
+          vacation.value.emergency_leaves = emergency_leaves.all
+        }
+      }
+    })
+
+    onMounted(async () => {
+      officeUsers.value = await $api.users.admin.office_users.list()
```
Comment (MohamedElmdary) [inline client/src/components/SetUserVacations.vue]: Handle Error and Loading statuses

File: client/src/components/SetUserVacations.vue [hunk 6]
```diff
@@ -0,0 +1,107 @@
...
+      users.value = officeUsers.value?.map((user: any) => ({ id: user.id, name: user.full_name }))
+      selectedUser.value = users.value[0]
+    })
+
+    async function updateVacations() {
+      await $api.vacations.balance.update({ user_ids: selectedUser.value.id }, vacation.value)
```
Comment (MohamedElmdary) [inline client/src/components/SetUserVacations.vue:99]: Handle Error and Loading statuses

File: client/src/components/SetVacations.vue [hunk 7]
```diff
@@ -0,0 +1,144 @@
...
+      const combinedDates = [...selectedDates.value]
+
+      return combinedDates
+        .sort((a, b) => a.getTime() - b.getTime())
+        .map((date) => formatDate(date))
+        .join(', ')
+    })
+
+    onMounted(async () => {
+      office_balance.value = await $api.vacations.get_admin_balance.list()
```
Comment (MohamedElmdary) [inline client/src/components/SetVacations.vue]: Handle Error and Loading statuses

File: client/src/components/SetVacations.vue [hunk 8]
```diff
@@ -0,0 +1,144 @@
...
+      selectedDates.value = [
+        ...office_balance.value.public_holidays.map((date: any) => new Date(date))
+      ]
+    })
+
+    async function setVacations() {
+      const { annual_leaves, compensation, emergency_leaves, leave_excuses, year } =
+        office_balance.value
+      await $api.vacations.post_admin_balance.create({
```
Comment (MohamedElmdary) [inline client/src/components/SetVacations.vue]: Handle Error and Loading statuses

File: client/src/components/SetVacations.vue [hunk 9]
```diff
@@ -0,0 +1,144 @@
...
+            <template v-slot:append-inner>
+              <v-icon>mdi-calendar</v-icon>
+            </template>
+          </v-text-field>
+
+          <v-date-picker
+            v-if="datePickerVisible"
+            v-model="selectedDates"
+            @click:prev="preventDefault"
+            @click:next="preventDefault"
```
Comment (MohamedElmdary) [inline client/src/components/SetVacations.vue]: use .prevent instead

```suggestion
            @click:prev.prevent
            @click:next.prevent
```

File: client/src/components/UpdateOfficeVacations.vue [hunk 10]
```diff
@@ -0,0 +1,90 @@
...
+    const selectedReason = ref(reasons.value[0])
+
+    const vacation = ref({
+      officeId: state.user.value.location.id,
+      value: 1,
+      reason: selectedReason.value.value
+    })
+
+    async function UpdateVacations() {
+      await $api.vacations.balance.adjustment.update(vacation.value)
```
Comment (MohamedElmdary) [inline client/src/components/UpdateOfficeVacations.vue:77]: Handle Error and Loading statuses

File: client/src/components/UpdateUser.vue [hunk 11]
```diff
@@ -0,0 +1,297 @@
...
+    const officeUsers = ref([])
+    const selectedUser = ref()
+    const reporting_to = ref()
+
+    onMounted(async () => {
+      offices.value = (await $api.office.list()).map((office: any) => ({
+        id: office.id,
+        name: office.name
+      }))
+      officeUsers.value = await $api.users.admin.office_users.list()
```
Comment (MohamedElmdary) [inline client/src/components/UpdateUser.vue]: Handle Error and Loading statuses

File: client/src/components/UpdateUser.vue [hunk 12]
```diff
@@ -0,0 +1,297 @@
...
+    const offices = ref([])
+    const supervisors = ref([])
+    const image = ref()
+    const imageUrl = ref()
+    const officeUsers = ref([])
+    const selectedUser = ref()
+    const reporting_to = ref()
+
+    onMounted(async () => {
+      offices.value = (await $api.office.list()).map((office: any) => ({
```
Comment (MohamedElmdary) [inline client/src/components/UpdateUser.vue]: Handle Error and Loading statuses

File: client/src/components/UpdateUser.vue [hunk 13]
```diff
@@ -0,0 +1,297 @@
...
+        id: office.id,
+        name: office.name
+      }))
+      officeUsers.value = await $api.users.admin.office_users.list()
+      selectedUser.value = officeUsers.value[0]
+      reporting_to.value = selectedUser.value.reporting_to[0]
+      supervisors.value = ((await $api.users.admin.list()) as any).filter(
```
Comment (MohamedElmdary) [inline client/src/components/UpdateUser.vue]: Handle Error and Loading statuses

File: client/src/utils/countries.ts
Comment (MohamedElmdary) [inline client/src/utils/countries.ts:1]: Why do we handle countries on our own?
can we use `country-look-up` package?

File: client/src/views/DashboardView.vue [hunk 14]
```diff
@@ -1,10 +1,84 @@
...
+        case 6:
+          return UpdateUser
+        default:
+          return SetVacations
+      }
+    })
+
+    onMounted(async () => {
+      officeId.value = state.user.value?.location.id
+      office.value = (await $api.office.read(officeId.value)).name
```
Comment (MohamedElmdary) [inline client/src/views/DashboardView.vue]: Handle Error and Loading statuses

File: client/src/views/LoginView.vue [hunk 15]
```diff
@@ -76,30 +76,35 @@
...
       await $api.auth.login(
         {
           email,
           password,
         },
-        rememberMe.value
-      );
-      $router.push("/");
+      )
+      const user = await $api.myprofile.getUser()
```
Comment (MohamedElmdary) [inline client/src/views/LoginView.vue]: Handle Error and Loading statuses

File: client/src/views/LoginView.vue [hunk 16]
```diff
@@ -76,30 +76,35 @@
...
+    const form = ref()
+    const email = ref<string>('')
+    const password = ref<string>('')
+    const show = ref<boolean>(false)
+    const rememberMe = ref<boolean>(false)
+    const $router = useRouter()
+    const state = useState()
+
     async function login(email: string, password: string) {
       await $api.auth.login(
```
Comment (MohamedElmdary) [inline client/src/views/LoginView.vue:100]: Handle Error and Loading statuses

File: client/src/views/ProfileView.vue [hunk 17]
```diff
@@ -4,77 +4,72 @@
...
 
-  
+    async function getProfile() {
+      user.value = state.user.value
     }
 
     async function getUserBalance() {
-      balance.value = await $api.vacations.getVacationBalance( { "user_ids": user.value?.id});
-
+      balance.value = await $api.vacations.getVacationBalance({ user_ids: user.value?.id })
```
Comment (MohamedElmdary) [inline client/src/views/ProfileView.vue]: Handle Error and Loading statuses

File: client/src/views/ProfileView.vue [hunk 18]
```diff
@@ -4,77 +4,72 @@
...
     }
 
     onMounted(async () => {
-      await getProfile();
-      await getUserBalance();
-
+      await getProfile()
+      await getUserBalance()
```
Comment (MohamedElmdary) [inline client/src/views/ProfileView.vue]: Handle Error and Loading statuses

File: client/src/views/TeamView.vue [hunk 19]
```diff
@@ -56,34 +64,33 @@ export default {
...
-    } 
+    function displayValue(value: any) {
+      return value ?? '-'
+    }
 
     onMounted(async () => {
-      team.value = await $api.users.team.list();
-      supervisors.value = await $api.users.team.supervisors.list();
+      team.value = await $api.users.team.list()
+      supervisors.value = await $api.users.team.supervisors.list()
```
Comment (MohamedElmdary) [inline client/src/views/TeamView.vue]: Handle Error and Loading statuses

File: client/src/utils/countries.ts
Comment (zaelgohary) [reply client/src/utils/countries.ts:1]: All I need is the country name. No more features are needed, so I figured I could use countries from locally exported object.

**Comment by MohamedElmdary** [comment]
image
Make input of `type="password"` and add confirm new password

**Review by MohamedElmdary — APPROVED** [review]

**Pull request merged**